STAR-782: Port cursor compaction

blambov opened the pull request:
This ports the cursor compaction code from DSE ([DB-4159](https://datastax.jira.com/browse/DB-4159)). The patch is a little more extensive than the DSE version:
- Also adds support for compacting to BIG format.
- Refactors writers to remove repetition and extract the common core.
- Fixes a problem where a separate static clustering was created for the `byte[]` accessor but code was using identity equality with `Clustering.STATIC_CLUSTERING`.
- Cursor compaction is enabled for all strategies.
- Column subsets are decoded more efficiently.
- Disables key cache migration during compaction as it is a performance bottleneck.

blambov commented:
Some quick benchmark results from my local machine:

Reading files through cursor vs iterator:
```
Benchmark                                                      (dataBuilder)  (overlapRatio)  (size)  (sstableCount)  Mode  Cnt    Score    Error  Units
CompactionBreakdownBenchmark.iterateThroughCursor                    DEFAULT             1.0       2               2  avgt   10   95.469 ±  6.265  ms/op
CompactionBreakdownBenchmark.iterateThroughTableScanner              DEFAULT             1.0       2               2  avgt   10  513.951 ± 22.563  ms/op
```
Merging two iterators through cursor vs iterator:
```
CompactionBreakdownBenchmark.iterateThroughMergeCursor               DEFAULT             1.0       2               2  avgt   10  145.297 ±  5.302  ms/op
CompactionBreakdownBenchmark.iterateThroughMergeIterator             DEFAULT             1.0       2               2  avgt   10  701.817 ± 27.792  ms/op
```

The above plus performing purging and empty partition/row removal for compaction:
```
CompactionBreakdownBenchmark.iterateThroughCompactionCursor          DEFAULT             1.0       2               2  avgt   10  175.830 ± 11.510  ms/op
CompactionBreakdownBenchmark.iterateThroughCompactionIterator        DEFAULT             1.0       2               2  avgt   10  799.199 ± 40.436  ms/op
```

blambov commented:
This patch is now ready for review.

@mfleming can do the review, but I would also like @jacek-lewandowski to take a look, especially at the writer refactoring which is most probably also needed for the SSTable API and OSS port of the BTI format.

jacek-lewandowski commented on src/java/org/apache/cassandra/io/sstable/compaction/BigTableCursor.java:
```diff
@@ -0,0 +1,441 @@
+ * Cursor over sstable data files.
+ * Supports both BIG and BTI formats (which differ only in index and whose data file formats are identical).
+ */
+public class BigTableCursor implements SSTableCursor
```
personally I'd prefer something more like `DefaultTableCursor` to avoid confusion

JeremiahDJordan commented on src/java/org/apache/cassandra/db/ArrayClustering.java:
```diff
@@ -32,7 +32,7 @@ public ArrayClustering(byte[]... values)
     public long unsharedHeapSize()
     {
-        if (this == ByteArrayAccessor.factory.clustering() || this == ByteArrayAccessor.factory.staticClustering())
+        if (this == ByteArrayAccessor.factory.clustering())
```
I assume this change is because of the removal of STATIC_CLUSTERING from ByteArrayObjectFactory?

blambov replied:
Yes, a byte array clustering can no longer be static (and the reason for that is that using a separate byte array static clustering causes serious problems because a lot of code compares with `STATIC_CLUSTERING` directly).

JeremiahDJordan commented on src/java/org/apache/cassandra/io/sstable/SSTableRewriter.java:
```diff
@@ -224,18 +225,7 @@ private void moveStarts(SSTableReader newReader, DecoratedKey lowerbound)
             return;
 
         newReader.setupOnline();
-        List<DecoratedKey> invalidateKeys = null;
```
why do we no longer need to invalidate keys?

blambov replied:
It seems I forgot a `break` in the migration code, and I'm rerunning the benchmark, also with a better configuration for iterator compaction so that migration does not put it over the edge.

If the results show an improvement in latency or a low enough overhead, I will bring back migration before committing.

blambov replied:
The migration tests are [ready](https://fallout.sjc.dsinternal.org/performance/branimir.lambov@datastax.com/report/45aa5681-4831-44b1-a936-12bbb577fc08).

I do not see any improvement in [latency](https://fallout.sjc.dsinternal.org/artifacts/performance_reports/branimir.lambov@datastax.com/45aa5681-4831-44b1-a936-12bbb577fc08/report.html?metric=Latency_99th&operation=latency_1_1%3Alatency_1_1.result-success&operation_filter=.%2Bresult-success&smoothing=1&show_aggregates=true&min_ops=0&xmin=0&xmax=5159.3&ymin=0&ymax=40) with cache migration enabled, while there is a very visible deterioration in [throughput](https://fallout.sjc.dsinternal.org/artifacts/performance_reports/branimir.lambov@datastax.com/45aa5681-4831-44b1-a936-12bbb577fc08/report.html?metric=Ops_Sec&operation=throughput%3Athroughput.result-success&operation_filter=.%2Bresult-success&smoothing=1&show_aggregates=true&min_ops=0&xmin=0&xmax=61149.63&ymin=0&ymax=364312.3) (250k -> 220k for cursors, 197k -> 180k for iterators) caused by significantly slowed down compaction.

To improve compaction performance and simplify configuration, I am going to remove the feature. Please let me know if you do not agree with this decision.

JeremiahDJordan reviewed (approved):
Reviewed this with an eye at comparing it with the DSE code.  A couple very minor questions but from that stand point of "does this do what the bdp code does" it looks good to me.

mfleming commented on src/java/org/apache/cassandra/utils/Reducer.java:
```diff
@@ -0,0 +1,94 @@
+    /**
+     * @return true if Out is the same as In for the case of a single source iterator
+     */
+    public boolean trivialReduceIsTrivial()
```
Could we change this to simply `reduceIsTrivial()`?

blambov replied:
Renamed to `singleSourceReduceIsTrivial()`.

mfleming reviewed (approved):
I left a few comments but there's nothing substantial in the review. Looks great!

blambov commented on src/java/org/apache/cassandra/db/compaction/writers/CompactionAwareWriter.java:
```diff
@@ -137,7 +139,32 @@ public long estimatedKeys()
+        return sstableWriter.append(partition);
+    }
+
+    public boolean startPartition(DecoratedKey partitionKey, DeletionTime deletionTime) throws IOException
```
Done

mfleming commented on src/java/org/apache/cassandra/db/rows/UnfilteredRowIteratorWithLowerBound.java:
```diff
@@ -76,10 +84,17 @@ public UnfilteredRowIteratorWithLowerBound(DecoratedKey partitionKey,
+        state = State.LOWER_BOUND_NOT_REQUESTED;
+    }
+
+    public void produceLowerBound()
```
Yeah, `request` sounds good.

blambov commented:
Squashed and rebased in preparation to commit. The individual commits can be found in [blambov:STAR-782-pre-squash](https://github.com/blambov/cassandra/commits/STAR-782-pre-squash).
